overlay.d/25rhcos-azure-udev/usr/lib/udev/rules.d/80-azure-disk.rules

```
@@ -0,0 +1,50 @@
+ATTRS{device_id}=="?00000000-0000-*", ENV{AZURE_DISK_TYPE}="os", GOTO="azure_disk_symlink"
+ENV{DEVTYPE}=="partition", PROGRAM="/bin/sh -c 'readlink /sys/class/block/%k/../device|cut -d: -f4'", ENV{AZURE_DISK_LUN}="$result"
+ENV{DEVTYPE}=="disk", PROGRAM="/bin/sh -c 'readlink /sys/class/block/%k/device|cut -d: -f4'", ENV{AZURE_DISK_LUN}="$result"
+ATTRS{device_id}=="{f8b3781a-1e82-4818-a1c3-63d806ec15bb}", ENV{AZURE_DISK_LUN}=="0", ENV{AZURE_DISK_TYPE}="os", ENV{AZURE_DISK_LUN}="", GOTO="azure_disk_symlink"
+ATTRS{device_id}=="{f8b3781b-1e82-4818-a1c3-63d806ec15bb}", ENV{AZURE_DISK_TYPE}="data", GOTO="azure_disk_symlink"
+ATTRS{device_id}=="{f8b3781c-1e82-4818-a1c3-63d806ec15bb}", ENV{AZURE_DISK_TYPE}="data", GOTO="azure_disk_symlink"
+ATTRS{device_id}=="{f8b3781d-1e82-4818-a1c3-63d806ec15bb}", ENV{AZURE_DISK_TYPE}="data", GOTO="azure_disk_symlink"
+
+# Use "resource" type for local SCSI because some VM skus offer NVMe local disks in addition to a SCSI resource disk, e.g. LSv3 family.
+# This logic is already in walinuxagent rules but we duplicate it here to avoid an unnecessary dependency for anyone requiring it.
+ATTRS{device_id}=="?00000000-0001-*", ENV{AZURE_DISK_TYPE}="resource", ENV{AZURE_DISK_LUN}="", GOTO="azure_disk_symlink"
+ATTRS{device_id}=="{f8b3781a-1e82-4818-a1c3-63d806ec15bb}", ENV{AZURE_DISK_LUN}=="1", ENV{AZURE_DISK_TYPE}="resource", ENV{AZURE_DISK_LUN}="", GOTO="azure_disk_symlink"
+GOTO="azure_disk_end"
+
+LABEL="azure_disk_nvme_direct_v1"
+LABEL="azure_disk_nvme_direct_v2"
+ATTRS{nsid}=="?*", ENV{AZURE_DISK_TYPE}="local", ENV{AZURE_DISK_SERIAL}="$env{ID_SERIAL_SHORT}"
+GOTO="azure_disk_nvme_id"
+
+LABEL="azure_disk_nvme_remote_v1"
+ATTRS{nsid}=="1", ENV{AZURE_DISK_TYPE}="os", GOTO="azure_disk_nvme_id"
+ATTRS{nsid}=="?*", ENV{AZURE_DISK_TYPE}="data", PROGRAM="/bin/sh -ec 'echo $$((%s{nsid}-2))'", ENV{AZURE_DISK_LUN}="$result"
+
+LABEL="azure_disk_nvme_id"
+ATTRS{nsid}=="?*", IMPORT{program}="/usr/sbin/azure-nvme-id --udev"
```

There is a diff from upstream (fedora RPM) here:

```patch 
--- /usr/lib/udev/rules.d/80-azure-disk.rules   1970-01-01 00:00:00.000000000 +0000
+++ /var/mnt/workdir/80-azure-disk.rules        2025-03-20 16:12:18.020106392 +0000
@@ -32,7 +32,7 @@
 ATTRS{nsid}=="?*", ENV{AZURE_DISK_TYPE}="data", PROGRAM="/bin/sh -ec 'echo $$((%s{nsid}-2))'", ENV{AZURE_DISK_LUN}="$result"
 
 LABEL="azure_disk_nvme_id"
-ATTRS{nsid}=="?*", IMPORT{program}="/usr/bin/azure-nvme-id --udev"
+ATTRS{nsid}=="?*", IMPORT{program}="/usr/sbin/azure-nvme-id --udev"
 
 LABEL="azure_disk_symlink"
 # systemd v254 ships an updated 60-persistent-storage.rules that would allow
```

Though.. this really highlights a problem here that I didn't understand before. Previously the udev rules we provided were just rules and didn't require external software.

Here these rules are running `/usr/sbin/azure-nvme-id`, but we aren't shipping that here (because we aren't shipping an RPM).

I think this may be a blocker for us for el9.

---

This is interesting because these rules are directly copied from the `azure-vm-utils` github repo, so there's a diff between that file and the udev rules shipped in the RPM.

as for the **azure-nvme-id** binary: The tool's [description](https://github.com/Azure/azure-vm-utils/blob/9c596916b6774f24420dac0ee7a72a6c9ddb5060/doc/azure-nvme-id.md#description) says:
```
azure-nvme-id provides identification metadata in the response to Identify Namespace command
for some models of NVMe devices. This is found in vendor-specific (vs) field which contains various
identification details with a comma-separated, key=value format.
```

I'm not sure what, within Azure, would rely on this vendor specific data, but there could be unexpected behavior if the binary is missing. However, the original bug was confirmed to be resolved by adding these udev rules without installing the whole RPM, and using them in an RHCOS Azure instance seems to work as expected.

---

> This is interesting because these rules are directly copied from the `azure-vm-utils` github repo, so there's a diff between that file and the udev rules shipped in the RPM.

yeah. In rawhide there was a usr/bin usr/sbin migration thing and that probably made its way into f41/f42 RPMs too so that probably is what this is about. It probably shouldn't matter too much I don't guess.





> I'm not sure what, within Azure, would rely on this vendor specific data, but there could be unexpected behavior if the binary is missing. However, the original bug was confirmed to be resolved by adding these udev rules without installing the whole RPM, and using them in an RHCOS Azure instance seems to work as expected.

So I think you're telling me the rules we need aren't this one that runs this binary?

I guess it's OK then, but I'd almost rather remove the lines that rely on the binary (or comment them out and add a comment why).

---

Yes, the other rules still work and set everything up correctly even if the binary is missing. 

```suggestion
#ATTRS{nsid}=="?*", IMPORT{program}="/usr/sbin/azure-nvme-id --udev"
```
I commented out the line, but left `LABEL="azure_disk_nvme_id"` intact so the other rules with `GOTO="azure_disk_nvme_id"` continue to function properly. The behavior was the same with or without that line commented out:

- `AZURE_DISK_LUN` is set to a value on NVMe data disks
- data and os NVMe disks show `AZURE_DISK_TYPE=data` and `AZURE_DISK_TYPE=os` respectively 
- NVMe disks and their partitions show up in `/dev/disk/azure` and `/dev/disk/azure/data/by-lun` as expected

This means the udev rule that calls the binary must be failing silently and isn't necessary for this specific bug fix.

---

I'm happy to have it commented out here in the code with a comment that says why we commented it out and then we should be able to move forward I think.

---

done!